xarray/core/datatree.py

```diff
@@ -1486,6 +1490,50 @@ def groups(self):
         """Return all groups in the tree, given as a tuple of path-like strings."""
         return tuple(node.path for node in self.subtree)
 
+    def _unary_op(self, f, *args, **kwargs) -> DataTree:
+        # TODO do we need to any additional work to avoid duplication etc.? (Similar to aggregations)
+        return self.map_over_subtree(f, *args, **kwargs)  # type: ignore[return-value]
+
+    def _binary_op(self, other, f, reflexive=False, join=None) -> DataTree:
+        from xarray.core.dataset import Dataset
+        from xarray.core.groupby import GroupBy
+
+        if isinstance(other, GroupBy):
+            # TODO should we be trying to make this work?
+            raise NotImplementedError
+
+        ds_binop = functools.partial(
+            Dataset._binary_op,
+            f=f,
+            reflexive=reflexive,
+            join=join,
+        )
+        return map_over_subtree(ds_binop)(self, other)
+
+    def _inplace_binary_op(self, other, f) -> Self:
+        from xarray.core.groupby import GroupBy
+
+        if isinstance(other, GroupBy):
+            raise TypeError(
+                "in-place operations between a DataTree and "
+                "a grouped object are not permitted"
+            )
+
+        # TODO requires an implementation of map_over_subtree_inplace
```

Comment:
I suspect this actually needs a different implementation, in order to handle error recovery properly.

I would suggest handling this like `Dataset._inplace_binary_op`, which builds result `Dataset` and then replaces the contents of the current dataset with those of the new dataset.

Reply:
Makes sense. I've raised https://github.com/pydata/xarray/issues/9629 to track that, so I'll remove this pseudocode from this PR.

Reply:
done in https://github.com/pydata/xarray/pull/9619/commits/304eb19a6cb427bc84da771c87f3780bc94e0356